elements/pl-matching/pl-matching.py

```diff
@@ -0,0 +1,399 @@
+                    else:
+                        html_params['incorrect'] = True
+                except Exception:
+                    raise ValueError('invalid score' + score)
+
+            with open('pl-matching.mustache', 'r', encoding='utf-8') as f:
+                html = chevron.render(f, html_params).strip()
+    elif data['panel'] == 'answer':
+        if not pl.get_boolean_attrib(element, 'hide-answer-panel', HIDE_ANSWER_PANEL_DEFAULT):
+            correct_answer_list = data['correct_answers'].get(name, [])
+
+            answerset = []
+            for i, answer in enumerate(display_answers):
+                form_name = get_form_name(name, answer['key'])
+                correct_answer = correct_answer_list[i]
+
+                answer_html = {
+                    'html': answer['html'].strip(),
+                    'options': [{'value': get_counter(correct_answer, counter_type)}],
+                }
+                answerset.append(answer_html)
+
+            optionset = []
+            for option in display_options:
+                option_html = {
+                    'key': option['key'],
+                    'html': option['html'].strip()
+                }
+                optionset.append(option_html)
+
+            html_params = {
+                'answer': True,
+                'answers': answerset,
+                'options': optionset,
+                'hide_letter_keys': pl.get_boolean_attrib(element, 'hide-letter-keys', HIDE_LETTER_KEYS_DEFAULT),
+                'counter_type': counter_type,
+            }
+            with open('pl-matching.mustache', 'r', encoding='utf-8') as f:
+                html = chevron.render(f, html_params).strip()
+
+    return html
+
+
+def grade(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, 'answers-name')
+    weight = pl.get_integer_attrib(element, 'weight', WEIGHT_DEFAULT)
+    partial_credit = pl.get_boolean_attrib(element, 'partial-credit', PARTIAL_CREDIT_DEFAULT)
+    counter_type = pl.get_string_attrib(element, 'counter-type', COUNTER_TYPE_DEFAULT)
+    display_answers, _ = data['params'][name]
+    number_answers = len(display_answers)
+
+    submitted_answers = data['submitted_answers']
+    correct_answers = data['correct_answers'].get(name, [])
+
+    # Count the number of answers that are correct.
+    num_correct = 0
+    for i in range(number_answers):
+        expected_html_name = get_form_name(name, i)
+        student_answer = submitted_answers.get(expected_html_name, None)
+        correct_answer = get_counter(correct_answers[i], counter_type)
+        if student_answer == correct_answer:
+            num_correct += 1
+
+    score = 0
+    if not partial_credit and num_correct == number_answers:
+        score = 1
+    elif partial_credit:
+        # EDC grading
+        number_wrong = number_answers - num_correct
+        score = 1 - 1.0 * number_wrong / number_answers
```

Comment:
Why not `num_correct / number_answers`? Note that in Python `/` is floating point division and `//` is integer division, so the `1.0 *` is not necessary.

Reply:
Copied this from pl-checkbox and turned off my brain 😳.